### PR TITLE
verify that IOThreadPools get shut down

### PR DESCRIPTION
**Pull request by weissi**
# DO NOT MERGE BEFORE 1.10.0 GOT RELEASED

This should go into 1.11.0

Motivation:

Not shutting down IOThreadPools leads to thread leaks so we should
verify they're actually shut down.

Modifications:

add a check that we're shutting down

Result:

fewer thread leaks

### CODE REVIEW

File: Sources/NIO/BlockingIOThreadPool.swift [hunk 1]
```diff
@@ -167,6 +167,15 @@ public final class BlockingIOThreadPool {
             }
         }
     }
+    
+    deinit {
+        switch self.state {
+        case .stopped, .shuttingDown:
+            ()
+        default:
+            assertionFailure("wrong state \(self.state)")
```
Comment (Lukasa) [inline Sources/NIO/BlockingIOThreadPool.swift]: What's the reasoning for this being an assertion instead of a precondition?

Comment (weissi) [reply Sources/NIO/BlockingIOThreadPool.swift]: @Lukasa same as for `BaseSocketChannel`, also it's not really a correctness issue if it's leaked I thought...

Comment (Lukasa) [reply Sources/NIO/BlockingIOThreadPool.swift]: Fair enough.

**Comment by Lukasa** [comment]
Looks like an integration test failure.

**Comment by weissi** [comment]
@Lukasa intermittent failure:

```
Fatal error: Posix wrapper adds more than 20% overhead (with wrapper: 0.806146025657654, without: 0.64440906047821
```

deffo unrelated

**Comment by weissi** [comment]
@swift-nio-bot test this please

**Pull request merged**